File: 백준/Gold/2098. 외판원 순회/외판원 순회.py
import sys

n = int(sys.stdin.readline())
inf = 16_000_001
cost = []
for _ in range(n):
    cost.append(list(map(int, sys.stdin.readline().split())))

info = [[-1] * (1<<n) for _ in range(n)] # info[i][mask] = mask 경로를 따라 i에 왔을 때, 그 이후 비용의 최솟값

def search(city, route):
    if route == (1 << n) - 1: return cost[city][0] if cost[city][0] > 0 else inf
    if info[city][route] != -1 : return info[city][route] # 이미 탐색함
    
    result = inf
    for i in range(n):
        if (1 << i) & route != 0 or cost[city][i] == 0: continue
        
        result = min(result, search(i, route | (1 << i)) + cost[city][i])
    
    info[city][route] = result
    return result

print(search(0, 1))

File: 백준/Gold/2098. 외판원 순회/외판원 순회.cc
#include <iostream>
#include <cstring>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
#define INF 987654321

using namespace std;

int n;
int cost[16][16];
int info[16][1<<16];

int search(int city, int route){
    if(route == (1 << n) - 1){
        if(cost[city][0] == 0) return INF;
        return cost[city][0];
    }
    if(info[city][route] != -1) return info[city][route];
    
    info[city][route] = INF;
    for(int i = 0; i < n; i++){
        if(((1 << i) & route) != 0 || cost[city][i] == 0) continue;
        info[city][route] = min(info[city][route], search(i, route | (1 << i)) + cost[city][i]);
    }
    
    return info[city][route];
}

int main(){
    FASTIO;
    
    cin >> n;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            cin >> cost[i][j];
        }
    }
    
    memset(info, -1, sizeof(info));
    
    cout << search(0, 1) << endl;
    
    return 0;
}